app/models/user.rb

```diff
@@ -25,80 +30,204 @@ def to_s
   alias_method :is_admin?,   :is_admin
-  alias_method :is_viewer?,  :is_viewer
   alias_method :is_manager?, :is_manager
+  alias_method :is_viewer?,  :is_viewer
 
   def login
     webauth ? webauth.login : sunetid
   end
 
-  ##
+  # Notes on definitions for known roles:
+  #
+  # At present, there are a few methods that all do the same thing;
+  # see all the `can_manage_*` methods in Dor::Governable, i.e.
+  # https://github.com/sul-dlss/dor-services/blob/develop/lib/dor/models/governable.rb
+  #
+  # The upshot of this is that it’s not yet clear how Argo should query a Dor
+  # object about management rights. That is, it’s not clear how granular the
+  # management privilege should be. It seems reasonable, at present, to simply
+  # query a Dor object using `dor_object.can_manage_item?` (without recourse to
+  # `dor_object.can_manage_content?` etc.)
+  #
+  # If more granular management rights are implemented, this method must evolve
+  # with them (possibly adding an additional parameter for the type of access
+  # requested).  Also, the ArgoHelper#render_buttons should be revised.
+
+  # Validate access permissions using Argo user roles. If that fails, check the
+  # APO role permissions.  If the object is an APO, first check permissions on
+  # the governing APO of this APO object. As a last resort, check the object
+  # roles when this object is an APO.
+  #
+  # @param dor_object [String|Dor::Base] Accepts a DRUID string or
+  #                                      any subclass of Dor::Base
+  # @return [Boolean]
+  def can_manage_object?(dor_object)
+    dor_object = get_dor_object(dor_object)
+    pid = dor_object.pid
+    apo_pid = dor_object.admin_policy_object.pid
+    manager_access = is_admin || is_manager
+    # Failing that, the user must belong to a workgroup permitted to manage the
+    # object.  The authorization is first checked in the object APO (apo_pid).
+    manager_access ||= dor_object.can_manage_item?(roles(apo_pid))
+    # Failing that, if the object is an APO, it can be the source of it's own
+    # authorization (pid).
+    manager_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_manage_item?(roles(pid))
+    manager_access
+  end
+
+  # Validate access permissions using Argo user roles. If that fails, check the
+  # APO role permissions.  If the object is an APO, first check permissions on
+  # the governing APO of this APO object. As a last resort, check the object
+  # roles when this object is an APO.
+  #
+  # @param dor_object [String|Dor::Base] Accepts a DRUID string or
+  #                                      any subclass of Dor::Base
+  # @return [Boolean]
+  def can_view_object?(dor_object)
+    dor_object = get_dor_object(dor_object)
+    # Any object administrator or manager can also view the object.
+    return true if can_manage_object?(dor_object)
+    pid = dor_object.pid
+    apo_pid = dor_object.admin_policy_object.pid
```

Comment:
maybe:
`apo_pid = dor_object.admin_policy_object.pid rescue nil`

in case the object in question doesn't have an APO, as with the current incarnation of `CatalogController#valid_user?`